### PR TITLE
SI-6380 Add @throws[Exception]

### PR DESCRIPTION
**Pull request by soc**
Compiler support.

### CODE REVIEW

**Comment by soc** [comment]
Review by @xeno-by, @paulp.

**Comment by gkossakowski** [comment]
Insufficient commit message.

**Comment by soc** [comment]
Thanks, I'll fix it.

**Comment by soc** [comment]
PLS REBUILD ALL!

File: src/reflect/scala/reflect/internal/AnnotationInfos.scala [hunk 1]
```diff
@@ -325,4 +325,23 @@ trait AnnotationInfos extends api.Annotations { self: SymbolTable =>
   implicit val AnnotationTag = ClassTag[AnnotationInfo](classOf[AnnotationInfo])
 
   object UnmappableAnnotation extends CompleteAnnotationInfo(NoType, Nil, Nil)
+  
+  /** Extracts symbol of thrown exception from AnnotationInfo.
+    * 
+    * Supports both “old-style” `@throws(classOf[Exception])`
+    * as well as “new-stye” `@throws[Exception]("cause")` annotations.
+    */
+  object ThrownException {
```
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/AnnotationInfos.scala:334]: Why not name it `Throws` to be consistent with the name of the annotation?

File: src/reflect/scala/reflect/internal/AnnotationInfos.scala [hunk 2]
```diff
@@ -325,4 +325,23 @@ trait AnnotationInfos extends api.Annotations { self: SymbolTable =>
...
+    def unapply(ann: AnnotationInfo): Option[Symbol] = 
+      ann match {
+        case AnnotationInfo(tpe, _, _) if tpe.typeSymbol != ThrowsClass =>
+          None
+        // old-style: @throws(classOf[Exception]) (which is throws[T](classOf[Exception]))
+        case AnnotationInfo(_, List(Literal(Constant(tpe: Type))), _) =>
+          Some(tpe.typeSymbol)
+        // new-style: @throws[Exception], @throws[Exception]("cause")
+        case AnnotationInfo(TypeRef(_, _, args), _, _) =>
+          Some(args.head.typeSymbol)
```
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/AnnotationInfos.scala:344]: I guess this one could be rewritten as:

```
case AnnotationInfo(TypeRef(_, _, List(arg)), _, _) =>
  Some(arg.typeSymbol)
```

Comment (paulp) [inline src/reflect/scala/reflect/internal/AnnotationInfos.scala:344]: Better yet (more efficient) and more resilient is

```
case AnnotationInfo(TypeRef(_, _, arg :: _), _, _) => Some(arg.typeSymbol)
```

File: test/files/run/t6380.scala [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  println(exs)
+}
+
+class Foo {
+  @throws[Exception]
+  def bar1 = ???
+  @throws[Throwable]("always")
+  def bar2 = ???
+  @throws(classOf[RuntimeException])
+  def bar3 = ???
```
Comment (retronym) [inline test/files/run/t6380.scala:11]: Could you add a test with multiple `@throws` annotations on one method?

File: src/reflect/scala/reflect/internal/AnnotationInfos.scala (hunk 1 above)
Comment (soc) [inline src/reflect/scala/reflect/internal/AnnotationInfos.scala:334]: Ok, I'll do that. I was a bit reluctant to claim this "straightforward" name.

File: test/files/run/t6380.scala (hunk 3 above)
Comment (soc) [inline test/files/run/t6380.scala:11]: Ok, I'll do that. I'll probably also need to check what happens when there are multiple throws annotations of the same type. What semantics would be reasonable here?

Comment (retronym) [inline test/files/run/t6380.scala:11]: `.distinct` seems reasonable.

Comment (soc) [inline test/files/run/t6380.scala:11]: Mhhh right! To be sure, I'll check the JVM spec, so we don't need to touch this again...

Comment (soc) [inline test/files/run/t6380.scala:11]: The JVM spec doesn't say anything on that matter ... it doesn't even mention whether it is supported/recommended to add non-checked exceptions to the Exceptions attribute.

Apart from that, I think we are good to go.

**Comment by gkossakowski** [comment]
PLEASE REBUILD ALL

**Comment by gkossakowski** [comment]
This has to be reviewed by Friday.

**Comment by paulp** [comment]
It looks reasonable to me.

**Comment by gkossakowski** [comment]
@paulp: Does it mean you stamped it with 'LGTM'? I want to be absolutely sure.

**Comment by paulp** [comment]
I was on my phone so I couldn't give it the most thorough endorsement, but I'll give it a closer look and let you know.

**Comment by paulp** [comment]
There are a couple minor suggestions which it would be good to see addressed, if @soc is around.  Failing that it LGTM anyway.

**Comment by gkossakowski** [comment]
Ok, thanks.

**Pull request merged**